### ISSUE TITLE
Exception ticking world - Server crash on village generation

### ISSUE THREAD

**Issue by Rutes** [SpongePowered/SpongeForge#2127]
**I am currently running**

- SpongeForge version: spongeforge-1.12.2-2655-7.1.0-BETA-3064
- Forge version: 2655
- Java version: 1.8.0_162
- Operating System: xubuntu 16.04

- Plugins/Mods:  
 Plugins (17): Minecraft, Minecraft Coder Pack, SpongeAPI, SpongeForge, EconomyLite, GriefPrevention, JdaLibraryLoader, LangSwitch, LuckPerms, MMCRestrict, Nucleus, ProjectCore, ProjectPortals, ServerListPlus, UltimateChat, VillagerShops, latch
 Mods (73): Minecraft, Minecraft Coder Pack, Forge Mod Loader, Minecraft Forge, SpongeAPI, SpongeForge, Angel of Vengeance, Animania, Antique Atlas, ArchitectureCraft, Aroma1997Core, AromaBackup, AromaBackuprecovery, Astral Sorcery, B.A.S.E, Baubles, BiblioCraft, Biomes O' Plenty, Chameleon, Clay WorldGen, Clumps, Corail Tombstone, CosmeticArmorReworked, CraftStudio API, Craftable Horse Armour [CHA&S], CreativeCore, CreativeCoreDummy, CustomNPCs, DungeonTactics, Dynamic Trees, Dynamic Trees BOP, Dynmap, Electroblob's Wizardry, EluCore, EluLib, Embers, Example Mod, Fairy Lights, Fast Leaf Decay, Guide-API, Ice and Fire, Inventory Tweaks, Iron Chest, Just Enough Items, Just a Few Fish, LLibrary, Less Forgiving Void, Level Up! Reloaded, Maille, Mantle, Morpheus, NetherEx, Player Revive, Roguelike Dungeons, Roots 2, Rustic, SimpleAutoRun, Skylands, Spartan Shields, Storage Drawers, SwingThroughGrass, TamModized, TeamLapen Library, Tinkers' Construct, ToroQuest, Totemic, Vampirism, VampirismIntegrations, Waila, Waystones, Wearable Backpacks, What Are We Looking At, WorldEdit

The 2 mods I have that definitely affect village generation are ToroQuest and Vampirism. I have a feeling that it is related to vampirism since the log references the "church".

During world exploration the server crashes upon encountering a village.

https://gist.github.com/Rutes/25ae239c2e0a6f0cddb43059db15b192

**Comment by Rutes** [comment]
pre-generating the world prevents the crash

**Comment by JBYoshi** [comment]
I don't think it's specifically part of village generation. The "church" is the [stone tower](https://minecraft.gamepedia.com/File:Villagechurch.png) generated in some villages.

It appears to be something related to villager professions/careers that are placed during village generation. Are there any mods you have that add professions or careers to villagers?

**Comment by JBYoshi** [comment]
Appears to be related to #2120.

**Comment by Rutes** [comment]
which part indicates the villager professions/careers? sorry I'm pretty new to troubleshooting crash logs. Is it net.minecraft.entity.passive.EntityVillager.func_175554_cu(EntityVillager.java:2268)?

**Comment by JBYoshi** [comment]
Yep. One way to analyze stack traces (without installing any extra programs) is by searching for bits of the stack trace in the [MCP mappings](http://www.modcoderpack.com) - download the bundle for the proper version of Minecraft and search the `conf/methods.csv` file for the bits of text starting with `func_`. In this case, `func_175554_cu` is listed as `populateBuyingList`. Typically, you only need to focus on a few listings in the crash report (after the last "Caused by:" line, if it exists).

If you ever plan on doing mod development, you can also use the "Analyze stack trace" feature of the [MinecraftDev plugin for IntelliJ IDEA](https://minecraftdev.org). It can clear them up all in one go, but it's somewhat overkill if you only need to work with a few crash reports at a time.

**Comment by gabizou** [comment]
I see the issue, I can "workaround" the issue by reporting it properly, but it means that sponge has not been able to validate the profession correctly because there are still awkward ways that villager careers are working with almost a lack of registrations. In either case, I understand what must be done, and then I can get the fix out soon.

In essence, [here](https://github.com/SpongePowered/SpongeForge/blob/bleeding/src/main/java/org/spongepowered/mod/mixin/core/entity/passive/MixinEntityVillager.java#L96) we need to catch if the careers are not synchronized and then force a resin of them on the villager registry to match.

**Comment by Rutes** [comment]
awesome! so, wait, is it a mod that is altering professions/careers? or just the plain old vanilla careers?

**Comment by gabizou** [comment]
It's more likely a mod career that isn't registered normally (there's currently several ways to register a career, which is the problem).

**Comment by Rutes** [comment]
Vampirism dev indicated this could be related to Vampire Hunters from that mod:
https://github.com/TeamLapen/Vampirism/issues/279#issuecomment-396016018

**Comment by maxanier** [comment]
I did not say that. The issue is unrelated to the hunter, but probably triggered by Vampirism in this case anyway: 
Vampirism creates a new VillageCareer for the `priest`. And based on gabizou's comment I guess we do not register it correctly (or at least not in a Sponge friendly way, it is working with Forge only)

@gabizou Do you have a example of a registration that is compatible with Sponge somewhere?
Vampirism does the registration during init here: https://github.com/TeamLapen/Vampirism/blob/029488905e2ba6d1a0efb57196378ba1534ef48a/src/main/java/de/teamlapen/vampirism/core/ModVillages.java#L76-L76

**Comment by gabizou** [comment]
I'm investigating the issue, since I've recently rewrote the registration, I could have missed some place.

**Comment by Y0ungSandwich** [comment]
Having the same issue on the modpack:
 Mech & Magic version: 3.0.2 
spongeforge version: spongeforge-1.12.2-2705-7.1.0-BETA-3169.

https://pastebin.com/rCu4Eif3

**Comment by gabizou** [comment]
Ok, once 3178 is finished building, try that build out. I'm trying to reproduce the issue and get villagers synced, but if this build fixes it, I'll close the issue.

**Comment by gabizou** [comment]
@Y0ungSandwich @Rutes give the new build a shot.

**Comment by Y0ungSandwich** [comment]
Loaded up fine with 3178 and I flew around several villages just to be sure. Everything is looking good!

**Comment by Rutes** [comment]
still crashing for me: https://gist.github.com/Rutes/e3809db85fc76ec9182d70bb43956660

**Comment by flamechair** [comment]
Crashing still for me as well, using 3178 and 3179

**Comment by gabizou** [comment]
Give 3181 a try. (it's being built as I write this).

Ok, new error this time. Correcting.

**Comment by flamechair** [comment]
Generating villages no longer causes the server/client to crash, instead this message is printed to the server log: https://gist.github.com/iamtpage/3bc95420535c1da65fc8b69441f3147d

**Comment by gabizou** [comment]
Alright, so, now try 3182 (building currently), I've tested it locally, but I wasn't able to get the same error you did @iamtpage. It's sounding like a separate issue, but hopefully the commit that resolved this issue should resolve yours.

Edit: Build uploaded.

**Comment by gabizou** [comment]
Verified the built jar is allowing snow villagers to spawn and be interacted correctly. @iamtpage If you still have issues with 3182 printing out that message, open a new issue.

**Comment by flamechair** [comment]
Same error/results as 3181: https://gist.github.com/iamtpage/eed7ae1e0047455781997467ff7e682e

**Comment by flamechair** [comment]
Sorry, didn't see the new comment. Will open a new issue

**Comment by Rutes** [comment]
confirmed no more crashes! Thanks guys!
